src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs

```diff
@@ -833,6 +833,62 @@ public void SocketSendReceiveBufferSize_SetZero_ThrowsSocketException()
                 Assert.Equal(e.SocketErrorCode, SocketError.InvalidArgument);
             }
         }
+
+        [Fact]
+        public void SendRecvIovMax_Success()
+        {
+            // sending/receiving more than IOV_MAX segments causes EMSGSIZE on some platforms.
+            // This is handled internally so this error shouldn't surface.
+            const int IOV_MAX = 1024; // Linux
+            using (var server = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                server.BindToAnonymousPort(IPAddress.Loopback);
+                server.Listen(1);
+
+                using (var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+                {
+                    client.Connect(server.LocalEndPoint);
+
+                    using (Socket acceptSocket = server.Accept())
+                    {
+                        // client: send data as segmentCount (> IOV_MAX) 1-byte segments.
+                        int segmentCount = IOV_MAX + 1;
+                        var sendBuffer = new byte[segmentCount];
+                        var sendSegments = new List<ArraySegment<byte>>();
+                        for (int i = 0; i < segmentCount; i++)
+                        {
+                            sendBuffer[i] = (byte)(i & 0xff);
+                            sendSegments.Add(new ArraySegment<byte>(sendBuffer, i, 1));
+                        }
+                        SocketError error;
+                        int bytesSent = client.Send(sendSegments, SocketFlags.None, out error);
```

Comment:
Couldn't this deadlock if the send buffer can't store all of the data being sent, since the receiver runs serially after it (especially if we oversee the number of segments)? Seems like they should be made to run in parallel.

Reply:
Sendbuffer/receivebuffer are much higher than what we are sending here, so this won't block. (my Fedora 25 defaults to 212992 buffers for send&receive)
I don't mind rewriting it to run in parallel though.